### ISSUE TITLE
react-router: `withRouter` fails with `ComponentType` starting in 3.6.2

### ISSUE THREAD

**Issue by OliverJAsh** [DefinitelyTyped/DefinitelyTyped#38271]
This worked in 3.5.2, but fails in 3.6.2.

`strict` mode is disabled.

Is this a bug in the typings or TypeScript?

I added failing tests for this here: https://github.com/DefinitelyTyped/DefinitelyTyped/commit/f036856dfc9a7fd140f378402c102206738ec0a4.

``` ts
import * as React from 'react';

// Copied from https://github.com/DefinitelyTyped/DefinitelyTyped/blob/5db94ca7fd3570dc23588b404d7bb669a7886a8a/types/react-router/index.d.ts#L146
declare function withRouter<P, C extends React.ComponentType<P>>(
    component: C & React.ComponentType<P>,
): unknown;

declare const Component: React.ComponentType<{}>;
/*
Argument of type 'ComponentType<{}>' is not assignable to parameter of type 'ComponentClass<{}, any> | (ComponentClass<{}, any> & FunctionComponent<{}>)'.
  Type 'FunctionComponent<{}>' is not assignable to type 'ComponentClass<{}, any> | (ComponentClass<{}, any> & FunctionComponent<{}>)'.
    Type 'FunctionComponent<{}>' is not assignable to type 'ComponentClass<{}, any> & FunctionComponent<{}>'.
      Type 'FunctionComponent<{}>' is not assignable to type 'ComponentClass<{}, any>'.
        Type 'FunctionComponent<{}>' provides no match for the signature 'new (props: {}, context?: any): Component<{}, any, any>'.
*/
withRouter(Component);
```

---

If you know how to fix the issue, make a pull request instead.

- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @sergey-buturlakin @mrk21 @vasek17 @ngbrown @awendland @KostyaEsmukov @johnnyreilly @LKay @DovydasNavickas @huy-nguyen @grmiade @DaIgeb @egorshulga @rraina @pret-a-porter @t49tran @8enSmith @wezleytsai @eps1lon @HipsterBrown

If you do not mention the authors the issue will be ignored.

**Comment by eps1lon** [comment]
I won't have time to dig into this. I think we should just remove any usage of `ComponentType` and use plain functions with explicit props with regard to `ref` and `children`. `ComponentType` was always a bit problematic because of the union. If it didn't create errors it produced very large unions with other hocs. Library consumers shouldn't touch statics anyway so loosing those shouldn't be an issue.

**Comment by basslagter** [comment]
Got the same issue. Already reported on the react-router repo (https://github.com/ReactTraining/react-router/issues/6906) but got redirected to here.

**Comment by bkoltai** [comment]
I'm still seeing issues even after this fix. Error message is something like
```
 Argument of type '({ myProp, history }: RouteComponentProps<{}, StaticContext, any> & { myProp: boolean; }) => Element' is not assignable to parameter of type 'ComponentType<RouteComponentProps<{}, StaticContext, any>>'.
  Type '({ myProp, history }: RouteComponentProps<{}, StaticContext, any> & { myProp: boolean; }) => Element' is not assignable to type 'FunctionComponent<RouteComponentProps<{}, StaticContext, any>>'.
    Types of parameters '__0' and 'props' are incompatible.
      Type 'PropsWithChildren<RouteComponentProps<{}, StaticContext, any>>' is not assignable to type 'RouteComponentProps<{}, StaticContext, any> & { myProp: boolean; }'.
        Property 'myProp' is missing in type 'RouteComponentProps<{}, StaticContext, any> & { children?: ReactNode; }' but required in type '{ myProp: boolean; }'.
```
with the component defined as
```typescript
import { RouteComponentProps, withRouter } from "react-router-dom";

interface Props extends RouteComponentProps {
  myProp: boolean;
}

const Component = ({ myProp, match }: Props) => null

export default withRouter(Component);
```

**Comment by artola** [comment]
@karol-majewski I am on TS 3.6.3 and the latest version `@types/react-router: 5.0.4` is breaking, while returning to former is OK.

```js
export function withRouter<P extends RouteComponentProps<any>, C extends React.ComponentType<P>>(
  component: C & React.ComponentType<P>,
): React.ComponentClass<Omit<P, keyof RouteComponentProps<any>> & WithRouterProps<C>> & WithRouterStatics<C>;
```

**Comment by eps1lon** [comment]
I encourage everyone to open new issues. If we fix your particular issue you have no way of being notified if you only comment in a closed thread. The typescript playground now allows configuring the ts version +  config and is able to download type definitions. Including a link in a new issue helps maintainers a lot. Thank you :pray: 

It seems like this only throws with `strictFunctionTypes` and I have a solution in mind. We'll see if it works out.

**Comment by OliverJAsh** [comment]
@sandersn Can we re-open this since the fix was reverted?

**Comment by OliverJAsh** [comment]
IIRC, @eps1lon discovered that this is due to a TypeScript bug: https://github.com/microsoft/TypeScript/issues/33490

Fix will come in: https://github.com/microsoft/TypeScript/pull/34607.

**Comment by karol-majewski** [comment]
Works fine with the nightly version. [Playground Link](https://www.typescriptlang.org/play/?ts=3.8.0-dev.20191030&ssl=16&ssc=17&pln=16&pc=16#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwBQdA9I3AMIRjBYAmcRJcABYwYYVAC5mAc2AxBAVwBGAOlwlGAESwFgAO1lYANgE8AKsbA9N2vQZPnL3RosMRFjAKzdFATgAsuMgA7ATcAMweQQAM3LgATBEAHImKflF+3EGKigBsOT7ByTnIiciMMBZYqIyUeDAAtMTyMFhQjHrcWAAeytzKMKgAxAAyAIx+OXSduIbIlHzyuvjAELpwAO6yghgQza0APAAKADRscN0tutzo2HXK7OCrWLowDlhHAHwfABR0cP9wNSPXTPGDiM4AMkwVBg9xIkBBLzen2OdAAlODFgBrXQQda6ehTLAzOZYQGrVDwB4I0Hg274OHA0HIgDeAF8PvRGAg6ABBKBSeQgUFwCAEOAVSzkalPJGVfbsj5kODAdC4+BoVDAKS6ZAuMkwCBwMBzZDClpQUXiyVksgyxEwVizVCoBVs07IXTGD5wAA+cG+9tBTs1bo9Xp9UIAYotlqsgy83R80WRlH84G9yDGljAVroEzAk8rVXB1SgXdrdfqJUabdL4bLHc7Xezw96-QGCyGXWGUBG4NHY7n4w2HUmU2mARnKlmh3mC0WVWqIBqKzq9YYDbWZ3bR8Hm73Pe3Bzn53vE4rU+mp5myNm4-nz4XL0vSyvy1r19XDRKd12D62fbeleU43ju97Do+TIXhyypgMQABuwCdMucAgMgMC4IIfDQBKghkp+uowPI8xkCC6wBvBHASHAgFqC8FwAPzgkeGJsE+h5em2SppggjB0Jscg7HsUCBk+aI0EAA)

**Comment by skylarmb** [comment]
bump on this?

**Comment by cristiantorresf19191919** [comment]
typescript is a waste of time, I have spent hours trying to solve this stupid error and nothins, so if typescript is made to make life easier it is definately not doing the job, full of errors and bugs

**Comment by ahejlsberg** [comment]
This issue was fixed in TS 3.7. Closing.
